account-compression/programs/account-compression/src/lib.rs

```diff
@@ -0,0 +1,498 @@
+//! SPL Account Compression also provides the ability to cache the upper most leaves of the
+//! concurrent merkle tree. This is called the "canopy", and is stored at the end of the
+//! ConcurrentMerkleTreeAccount. More information can be found in the initialization instruction
+//! documentation.
+//!
+//! While SPL ConcurrentMerkleTrees can generically store arbitrary information,
+//! one exemplified use-case is the [Bubblegum](https://github.com/metaplex-foundation/metaplex-program-library/tree/master/bubblegum) contract,
+//! which uses SPL-Compression to store encoded information about NFTs.
+//! The use of SPL-Compression within Bubblegum allows for:
+//! - up to 1 billion NFTs to be stored in a single account on-chain (>10,000x decrease in on-chain cost)
+//! - up to 2048 concurrent updates per slot
+//!
+//! Operationally, SPL ConcurrentMerkleTrees **must** be supplemented by off-chain indexers to cache information
+//! about leafs and to power an API that can supply up-to-date proofs to allow updates to the tree.
+//! All modifications to SPL ConcurrentMerkleTrees are settled on the Solana ledger via instructions against the SPL Compression contract.
+//! A production-ready indexer (Plerkle) can be found in the [Metaplex program library](https://github.com/metaplex-foundation/digital-asset-validator-plugin)
+
+use anchor_lang::{
+    emit,
+    prelude::*,
+    solana_program::sysvar::{clock::Clock, rent::Rent},
+};
+use borsh::{BorshDeserialize, BorshSerialize};
+use std::mem::size_of;
+
+pub mod canopy;
+pub mod data_wrapper;
+pub mod error;
+pub mod events;
+pub mod state;
+pub mod zero_copy;
+
+use crate::canopy::{fill_in_proof_from_canopy, update_canopy};
+use crate::data_wrapper::{wrap_event, Wrapper};
+use crate::error::AccountCompressionError;
+use crate::events::ChangeLogEvent;
+use crate::state::ConcurrentMerkleTreeHeader;
+use crate::zero_copy::ZeroCopy;
+
+/// Exported for Anchor / Solita
+pub use spl_concurrent_merkle_tree::{
+    concurrent_merkle_tree::ConcurrentMerkleTree, error::ConcurrentMerkleTreeError, node::Node,
+};
+
+declare_id!("GRoLLzvxpxxu2PGNJMMeZPyMxjAUH9pKqxGXV9DGiceU");
+
+/// Context for initializing a new SPL ConcurrentMerkleTree
+#[derive(Accounts)]
+pub struct Initialize<'info> {
+    #[account(zero)]
+    /// CHECK: This account will be zeroed out, and the size will be validated
+    pub merkle_tree: UncheckedAccount<'info>,
+
+    /// Authority that validates the content of the trees.
+    /// Typically a program, e.g., the Bubblegum contract validates that leaves are valid NFTs.
+    pub authority: Signer<'info>,
+
+    /// Program used to emit changelogs as instruction data.
+    /// See `WRAPYChf58WFCnyjXKJHtrPgzKXgHp6MD9aVDqJBbGh`
+    pub log_wrapper: Program<'info, Wrapper>,
+}
+
+/// Context for inserting, appending, or replacing a leaf in the tree
+///
+/// Modification instructions also require the proof to the leaf to be provided
+/// as base58-encoded nodes as "remaining accounts".
```

Comment:
nittiest of nits: if it's remaining accounts, it's not base58-encoded, just 32 bytes